lib/src/front_end/expression_contents.dart

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+
+/// Tracks the contents of a nested tree of argument lists and collection
+/// literals.
+///
+/// In general, the formatter tries to pack as much as it can on a single line
+/// until it hits the page width. However, with deeply nested call trees (which
+/// are pervasive in Flutter UI code), the expression nesting can get deep even
+/// in a short piece of code.
+///
+/// It can be much easier to track the nesting structure and identify siblings
+/// in the expression tree if it's forced to split more eagerly. Compare:
+///
+///     Apple(banana: [Cherry(date: Eggplant(1, 2))], fig: Grape(4))
+///
+///     Apple(
+///       banana: [
+///         Cherry(date: Eggplant(1, 2)),
+///       ],
+///       fig: Grape(4),
+///     )
+///
+/// This class records the necessary state to determine if a given collection
+/// literal or argument list is complex enough that it should be eagerly split.
+///
+/// It considers an operation A to contain another B if B occurs anywhere
+/// transitively inside the elements or argument list of A, regardless of any
+/// other AST nodes that may intercede. If we only looked at the immediate
+/// expressions in the collection or argument list to count nested calls and
+/// collections, then wrapping one of those expressions in, say, parentheses,
+/// could cause a nested operation to *not* be counted.
+///
+/// That would violate a reasonable principle that *adding* code to a call or
+/// collection should never cause it to go from splitting to not splitting. If a
+/// collection or call is complex enough to warrant splitting it eagerly, then
+/// adding more code in there should always lead to it still splitting.
+/// Tracking the contents transitively ensures that.
+///
+/// The heuristics for which collections and argument lists split are fairly
+/// simple and conservative and are documented below.
+class ExpressionContents {
+  /// The stack of calls and collections whose contents we are tracking and
+  /// that haven't completed yet.
+  final List<_Contents> _stack = [_Contents(_Type.otherCall)];
+
+  /// Begins tracking an argument list.
+  void beginCall(List<AstNode> arguments) {
+    var type = _Type.otherCall;
+    for (var argument in arguments) {
+      if (argument is NamedExpression) {
+        type = _Type.callWithNamedArgument;
+        if (!_isTrivial(argument.expression)) {
+          type = _Type.callWithNontrivialNamedArgument;
```

Comment:
[optional] Only this method that assigns `type` makes a differentiation between `callWithNamedArgument` and `callWithNontrivialNamedArgument`.

Consider collapsing those enums to `callWithNamedArgument` and tracking a separate boolean here.

```dart
    var hasNontrivialNamedArgument = false;
    for (var argument in arguments) {
      if (argument is NamedExpression) {
        type = _Type.callWithNamedArgument;
        if (!_isTrivial(argument.expression)) {
          hasNontrivialNamedArgument = true;
          break;
        }
      }
    }

    if (hasNontrivialNamedArgument) {
      _stack.last.nontrivialCalls++;
    }
```

Reply:
Good call. I didn't notice that we don't actually need to store that in `_Contents`. Thanks!